File: Data Structures & Algorithms/count-subsequences/submission-0.cpp
class Solution {

std::vector<std::vector<int>> dp;

public:
    int numDistinct(string s, string t) {
        dp.assign(s.length(), std::vector<int>(t.length(), -1));
        
        if (s.length() < t.length()) return 0;
        return dfs(0, 0, s, t);
    }

    int dfs(int i, int j, std::string& s, std::string& t) {
        if (j == t.length()) return 1;
        if (i == s.size()) return 0;

        if (dp[i][j] != -1) return dp[i][j];

        int res = dfs(i + 1, j, s, t);
        if (s[i] == t[j]) {
            res += dfs(i + 1, j + 1, s, t);
        }

        dp[i][j] = res;
        return dp[i][j];
    }
};

File: Data Structures & Algorithms/generate-parentheses/submission-0.py
class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        openb = 0
        closedb = 0
        res = []

        self.req_par(res, openb, closedb, n, curr_str='')

        return res
    
    def req_par(self, res: list, openb: int, closedb: int, n: int, curr_str: str):
        if closedb == n:
            print(closedb, openb)
            res.append(curr_str)

        elif openb == n:
            closedb += 1
            curr_str += ')'
            self.req_par(res, openb, closedb, n, curr_str)
        
        elif openb == closedb:
            openb += 1
            curr_str += '('
            self.req_par(res, openb, closedb, n, curr_str)
                    
        else:
            self.req_par(res, openb + 1, closedb, n, curr_str + '(')
            self.req_par(res, openb, closedb + 1, n, curr_str + ')')


        